bin/node/runtime/src/lib.rs

```diff
@@ -159,6 +159,7 @@ impl pallet_utility::Trait for Runtime {
 	type MultisigDepositBase = MultisigDepositBase;
 	type MultisigDepositFactor = MultisigDepositFactor;
 	type MaxSignatories = MaxSignatories;
+	type Dispatcher = frame_system::Module<Runtime>;
```

Comment:
Why do we need one `Dispatcher` for every module that needs to dispatch. Couldn't we just add it once to the `frame_system::Trait`?

Reply:
I didn’t put much thought into it TBH since it is the default approach. For example, we could have done the same with Call and others.


Having distinct types would allow customizing logic additionally. For example, you would be able to provide a dispatcher instance that would filter some calls or maybe add some logic. So maybe it is worth it to leave it that way?

Reply:
To make this possible `SignedExtension::post_dispatch` must be generic over the `Dispatcher` that belongs to the `Call` that we are post dispatching. We need this in order to bound it with something like `D: Dispatcher + WeightRefund`in order to get the spent weight from a generic Dispatcher. Right now in #5032 it hardcodes the system module there.

Reply:
What I meant not the total customization but rather adding some logic for the downstream modules, like `pallet-contracts`.

Reply:
Let's go with this example. `pallet_contracts` needs some additional logic in it's `Dispatcher`. Therefore the pallet adds some additional bounds onto it's `Dispatcher` type in it's pallet and the runtime impl supplies another `Dispatcher` as for other modules in order to fulfill the bound. This is then used for bookkeeping and all is fine.

**Until** some signed extension like the `ChargeTransactionPayment` needs to get an `Dispatcher` instance to get the unspent gas from. How would it determine which `Dispatcher` to query? It is only generic over the global overarching `Call`type. It does not generate code for every module.

For that reason in #5032 it just hardcodes the used dispatcher in `<ChargeTransactionPayment as SignedExtension>::post_dispatch`:
```
<frame_system::Module<T>>::spent_weight()
```

This won't work when the module that this extrinsic targets used another `Dispatcher` like in our example.

I therefore argue that we shouldn't allow a custom `Dispatcher` per module, because the point is that there is only one (specified by `frame_system::Trait::Dispatcher`.

Or would your custom `pallet_contract::Trait::Dispatcher` call into the `frame_system::Trait::Dispatcher`?

Reply:
Summary a discussion with @pepyakin is that it should be implemented in the following way: 

The `SignedExtension` will always call into the "main dispatcher" which is determined by `frame_system::Trait::Dispatcher`. When runtime implementors decide to use a different `Dispatcher` for some of their modules these "sub dispatchers" must always delegate to the "main dispatcher". The `SignedExtension` therefore only deals with one `Dispatcher` implementation and does not need to be generic in that regard.

This delegation contract also must be explained in the `Dispatcher` rustdoc.

Reply:
I implemented the discussed changes as a `MainDispatcher` type alias on the system trait to make it more obvious that it has a special meaning and is the one that custom dispatchers for individual modules should feed into.